### ISSUE TITLE
Send hot-reload signals

### ISSUE THREAD

**Issue by van800** [JetBrains/godot-support#3]
Like https://github.com/DavidKarlas/GodotExtension/blob/c3173c49db606bd35f9cc7687f8832442fe790c9/Godot/GodotDebuggerSession.cs#L19 sends
GD.Var2Bytes(new Array().Add("reload_scripts"))

Todo:
1. start listening to tcp $port
2. start game with additional argument `--remote-debug 127.0.0.1:$port`
3. After each build, send `reload_scripts` to that port. Use `com.jetbrains.rider.build.BuildHost` to receive event after build.

**Comment by van800** [comment]
Use messaging from https://github.com/JetBrains/godot-support/pull/102

**Comment by geowarin** [comment]
Hello @van800.
After having a look, I think sending a `ReloadScriptsRequest` from the messaging client is indeed a way to trigger a script reload.

I would like to try this approach, however, I'm not sure how to hook into the debbuger to trigger this request when the user recompiles his project.

I've never written a rider plugin before so any pointers would be welcome :smile:

**Comment by van800** [comment]
@geowarin why do you want to hook debugger, instead of the `BuildHost`?

**Comment by van800** [comment]
I see messages on the internet saying hot reload works in VS with some simple trick without calling any tcp signals.
https://www.reddit.com/r/godot/comments/16pjnt2/psa_c_hotreloading_works_with_this_one_simple/
So maybe we need to see VS and Rider side-by-side on windows to see what is different.

**Comment by geowarin** [comment]
> why do you want to hook debugger, instead of the `com.jetbrains.rider.build.BuildHost`?

I don't know, I thought this feature was only needed when running the game in debug mode?

Regarding, this message on reddit, I don't have VS to test but I really doubt that this works at all.
In theory, hot reloading should happen when the LSP sends a didSave request to the engine.

But this is not working. AFAIK hot reloading does not work with an external editor AT ALL.
See: https://github.com/godotengine/godot/issues/72825

**Comment by van800** [comment]
Here are my other thoughts:

1. When the game is started from the GodotEditor, we can use messaging from https://github.com/JetBrains/godot-support/pull/102 would to send signals to GodotEditor and it should be transmitted to the Game. However normally with Rider we start the Game directly from Rider, so we'd need to establish a separate tcp connection. This is likely blocked by the mentioned issue.

2. What we want is actually to execute this [3 lines of code](https://github.com/godotengine/godot/blob/5e512b705e66ecc86025e56bebd7632b8a392390/core/debugger/remote_debugger.cpp#L595-L597). Unfortunately, I can't find a way to do it. I can see the name of the script `Godot.ScriptLanguageExtension.MethodName._ReloadAllScripts`, but not the way to call it from the Game.

3. Hack way to accomplish the task could be to start Game with `--remote-debug 127.0.0.1:$port`, connect to it and send [this](https://github.com/DavidKarlas/GodotExtension/blob/c3173c49db606bd35f9cc7687f8832442fe790c9/Godot/GodotDebuggerSession.cs#L19) to its buffer. Here is an example of [SaveAll handler](https://github.com/JetBrains/resharper-unity/blob/3f57a1b9ce1894573605e7af96cac09d24f44ef6/rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/actions/RefreshInUnityAction.kt#L13) in Rider.

**Comment by geowarin** [comment]
From what I understand, calling the `ReloadScriptsRequest` from the IPC client introduced in the previous PR will call the 
[EditorDebuggerNode::reload_scripts()](https://github.com/godotengine/godot/blob/03598062fd259aa7198e65a56a157822b0720bc7/modules/mono/editor/editor_internal_calls.cpp#L150) method.

I think you're right, this might only reload the current tabs or something like that. I'll need to test that.
This makes me question if  `EditorDebuggerNode::reload_scripts()` which is callable from IPC should do something similar to the 3 lines you pointed out, from the remote debugger:

```
for (int i = 0; i < ScriptServer::get_language_count(); i++) {
    ScriptServer::get_language(i)->reload_all_scripts();
}
```

It would make sense to me at least.

The hacky way seems feasible also, let me give it a try.

**Comment by markusstephanides** [comment]
Hi, how is the progress on hot reloading gdscript? Would be very useful for us

**Comment by van800** [comment]
There is a technical problem on the Rider side https://youtrack.jetbrains.com/issue/RIDER-102927. Unfortunately, it is not being prioritized.

**Comment by markusstephanides** [comment]
Does this also apply to hot reloading gdscript? As only C# is mentioned in the issue you linked

**Comment by van800** [comment]
[RIDER-102927](https://youtrack.jetbrains.com/issue/RIDER-102927) is only about C#.
@markusstephanides if you can explain the gdscript reload to me, it would help a lot.

**Comment by markusstephanides** [comment]
If I change a .gdscript file in VS Code while the game is running (vs code didn't start the game, it is just connected to the Godot editor) and save the file, the changes will be immediately applied to the running game. I think this is also the case if you make changes to .gdscript files within the Godot editor.

**Comment by van800** [comment]
Thank you, I can confirm the missing feature in comparison with vscode. I will investigate further. Created a separate issue https://youtrack.jetbrains.com/issue/RIDER-114380/Godot-HotReload-GDScript. We use those youtrack issues to plan iterations.

**Comment by markusstephanides** [comment]
Thank you very much!

**Comment by van800** [comment]
Got this fixed, not sure when will it land. Follow the https://youtrack.jetbrains.com/issue/RIDER-114380/Godot-HotReload-GDScript

LSP was actually sending didSave signal, which should trigger the [reload](https://github.com/godotengine/godot/blob/0f5f3bc9546b46b2029fc8896dc859697f1eab97/modules/gdscript/language_server/gdscript_text_document.cpp#L109), but then unsupported signal (codeAction) was also send, which apparently broke the reload call. Also requested a more generic fix the in the Intellij platform https://youtrack.jetbrains.com/issue/IJPL-171159/codeActionProviderfalse-should-disable-the-codeAction-attempts
